ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -426,208 +432,228 @@ forkBlockForging ::
-              -- (we use the ticked ledger state). However, we probably don't
-              -- /want/ to produce a block in this case; we are most likely
-              -- missing a blocks on our chain.
-              trace $ TraceNoLedgerView currentSlot err
-              exitEarly
-            Right lv ->
-              return lv
-
-        trace $ TraceLedgerView currentSlot
-
-        -- Tick the 'ChainDepState' for the 'SlotNo' we're producing a block
-        -- for. We only need the ticked 'ChainDepState' to check the whether
-        -- we're a leader. This is much cheaper than ticking the entire
-        -- 'ExtLedgerState'.
-        let tickedChainDepState :: Ticked (ChainDepState (BlockProtocol blk))
-            tickedChainDepState =
-                tickChainDepState
-                  (configConsensus cfg)
-                  ledgerView
-                  currentSlot
-                  (headerStateChainDep (headerState unticked))
-
-        -- Check if we are the leader
-        proof <- do
-          shouldForge <- lift $
-            checkShouldForge blockForging
-              (contramap (TraceLabelCreds (forgeLabel blockForging))
-                (forgeStateInfoTracer tracers))
-              cfg
+      -- Get forker corresponding to bcPrevPoint
+      --
+      -- This might fail if, in between choosing 'bcPrevPoint' and this call to
+      -- 'ChainDB.getReadOnlyForkerAtPoint', we switched to a fork where 'bcPrevPoint'
+      -- is no longer on our chain. When that happens, we simply give up on the
+      -- chance to produce a block.
+      forkerEith <- lift $ ChainDB.getReadOnlyForkerAtPoint chainDB reg (SpecificPoint bcPrevPoint)
+      -- Remember to close this forker before exiting!
+      forker <- case forkerEith of
+        Left _ -> do
+          trace $ TraceNoLedgerState currentSlot bcPrevPoint
+          exitEarly
+        Right forker -> pure forker
+
+      unticked <- lift $ atomically $ LedgerDB.roforkerGetLedgerState forker
+
+      trace $ TraceLedgerState currentSlot bcPrevPoint
+
+      -- We require the ticked ledger view in order to construct the ticked
+      -- 'ChainDepState'.
+      ledgerView <-
+        case runExcept $ forecastFor
+                          (ledgerViewForecastAt
+                            (configLedger cfg)
+                            (ledgerState unticked))
+                          currentSlot of
+          Left err -> do
+            -- There are so many empty slots between the tip of our chain and the
+            -- current slot that we cannot get an ledger view anymore In
+            -- principle, this is no problem; we can still produce a block (we use
+            -- the ticked ledger state). However, we probably don't /want/ to
+            -- produce a block in this case; we are most likely missing a blocks
+            -- on our chain.
+            trace $ TraceNoLedgerView currentSlot err
+            lift $ roforkerClose forker
```

Comment:
Having to close it (and I suppose remove it from the registry) before every `exitEarly` call seems bad to me.

See my comment above on the `withRegistry` call. If my reasoning there is flawed, then perhaps we need a different monad to avoid this code smell.

Reply:
Good point. Maybe we *should* view calls to `exitEarly` as "exceptions" and let the resource registry handle the release actions, and then we only release at the end of the happy path

Reply:
#1316